src/compose/update-strategies.ts

```diff
@@ -12,12 +13,30 @@ export interface StrategyContext {
 	dependenciesMetForStart: boolean;
 	dependenciesMetForKill: boolean;
 	needsSpecialKill: boolean;
+	currentServices: Service[];
+	currentServicesLocked: boolean;
+	appsToLock: AppsToLockMap;
+}
+
+function generateLockThenKillStep(
+	current: Service,
+	currentServices: Service[],
+	currentServicesLocked: boolean,
+	appsToLock: AppsToLockMap,
+): Nullable<CompositionStep> {
+	if (!currentServicesLocked) {
+		currentServices.forEach((svc) =>
+			appsToLock[svc.appId].add(svc.serviceName),
+		);
+		return;
+	}
+	return generateStep('kill', { current });
 }
 
 export function getStepsFromStrategy(
 	strategy: string,
 	context: StrategyContext,
-): CompositionStep {
+): Nullable<CompositionStep> {
```

Comment:
If it makes sense to return `noop` here I think that's better. Returning `null` or "something" makes functions signatures more complicated and means that function callers now need to add a validation step

Reply:
I understand. In that case I would prefer that you changed the signature to return an array of next steps, and return an empty array if no further steps are necessary. Sometimes returning a null is not a good interface imho

Reply:
Done! Also in f199319d0a2ecf3dcbfc59c82f022b662d784a71